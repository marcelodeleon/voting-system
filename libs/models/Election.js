const mongoose = require('mongoose');

const electionSchema = new mongoose.Schema(
  {
    name: { type: String, required: true },
    description: { type: String, required: true },
    proposals: {
      type: {
        title: String,
        options: [String],
      },
      required: true,
    },
    startAt: { type: Date, required: true },
    endAt: { type: Date, required: true },
    //city: { type: String, required: true },
    //states: { type: String, required: true },
    //age: { type: String, required: true },
  },
  { timestamps: true },
);

const Election = mongoose.model('Election', electionSchema);

module.exports = Election;
