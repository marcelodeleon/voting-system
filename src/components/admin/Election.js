import React, { useState } from 'react';
import DatePicker from 'react-datepicker';
import ModalProposal from './ModalProposal';
import apiClient from '../../utils/api-client';

export default function Election() {
  const [name, setName] = useState('');
  const [description, setDescription] = useState('');
  const [startAt, setStartAt] = useState(new Date());
  const [endAt, setEndAt] = useState(new Date());
  const [isOpenModal, setIsOpenModal] = useState(false);
  let [proposalsList, setProposals] = useState([]);
  let proposalsObj = {};

  const loadProposals = (data) => {
    data.forEach((item) => {
      if (
        proposalsList.filter((elem) => elem.title === item.title).length === 0
      ) {
        setProposals((proposalsList = proposalsList.concat(item)));
      }
    });
    loadProposalsObj(proposalsList);
  };

  const loadProposalsObj = (data) => {
    const obj = {};
    for (let i = 0; i < data.length; i++) {
      obj[i] = data[i];
    }
    proposalsObj = obj;
  };

  const handleSubmit = async (evt) => {
    evt.preventDefault();
    try {
      await apiClient.post('createElections', {
        body: {
          electionData: {
            name,
            description,
            proposals: proposalsList,
            startAt,
            endAt,
          },
        },
      });
    } catch (error) {
      alert(error.message);
    }
  };

  const openModal = () => {
    setIsOpenModal(true);
  };

  const closeModal = () => {
    setIsOpenModal(false);
  };

  const cancelModal = (event) => {
    event.preventDefault();
    setIsOpenModal(false);
  };

  return (
    <div>
      <h1>Nueva Eleccion</h1>
      <a
        className="App-link"
        href="../.netlify/functions/hello"
        target="_blank"
        rel="noopener noreferrer"
      >
        Learn React
      </a>
      <form onSubmit={handleSubmit}>
        <label>
          Nombre:
          <input
            type="text"
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
        </label>
        <label>
          Descripcion:
          <input
            type="text"
            value={description}
            onChange={(e) => setDescription(e.target.value)}
          />
        </label>
        <label>
          Fecha Inicio:
          <DatePicker
            selected={startAt}
            onChange={(date) => setStartAt(date)}
          />
        </label>
        <label>
          Fecha Fin:
          <DatePicker selected={endAt} onChange={(date) => setEndAt(date)} />
        </label>
        <input type="submit" value="Submit" />
      </form>
      <button onClick={openModal}>Crear Propuestas</button>
      {isOpenModal && (
        <ModalProposal
          closeModal={closeModal}
          cancelModal={cancelModal}
          sendProposals={loadProposals}
        />
      )}
    </div>
  );
}
