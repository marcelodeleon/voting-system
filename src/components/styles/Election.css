.election-title {
    text-align: center;
    color: teal;
}

.election-form {
    padding: 1rem;
    border: 1px solid teal;
    border-radius: 1rem;
    width: 25rem;
    margin: auto;
}

.election-form label {
    font-weight: bold;
    padding-top: 1rem;
}

.election-form label,
.election-form input,
.election-form select {
    display: block;
    width: 20rem;
    margin: auto;
    color: teal;
}

.form-component {
    margin-top: 1rem !important;
    background: teal;
    color: white !important;
    font: inherit;
    border: 1.5px solid #0e4f1f;
    padding: 0.5rem;
    border-radius: 8px;
    font-weight: bold;
    cursor: pointer;
}

#root > div > form > input[type=submit]{
    display: block;
    margin-top: 1rem;
    height: 2.5rem;
}
