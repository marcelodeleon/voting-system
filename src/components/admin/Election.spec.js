import React from 'react';
import { render, fireEvent, waitFor, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import 'mutationobserver-shim';
import Election from './Election';
// import { isAuthenticated } from '../utils/auth';

global.MutationObserver = window.MutationObserver;
// jest.mock('../utils/auth');

describe('Election', () => {
  beforeEach(() => {
    // isAuthenticated.mockRestore()
  })

  test('Renders Election correctly', () => {
    const { container } = render(<Election />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test.only('test', async () => {
    render(<Election />);
    fireEvent.click(screen.getByText(/submit/i))

    // const test = await waitFor(() => screen.getByText(/error/i))

    console.log({ test })
  })
});
