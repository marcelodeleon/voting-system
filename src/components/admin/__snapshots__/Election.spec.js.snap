// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Election Renders Election correctly 1`] = `
<div>
  <div
    class="sc-bdfBwQ bWIvum Navbar"
  >
    <a
      class="Navbar__brand"
      href="/vote"
    >
      <img
        alt="logo"
        class="Navbar__brand-logo"
        src="logo.svg"
      />
    </a>
  </div>
  <h1
    class="election-title"
  >
    Nueva Eleccion
  </h1>
  <form
    class="election-form"
  >
    <label>
      Nombre:
      <input
        name="name"
        type="text"
        value=""
      />
    </label>
    <label>
      Descripcion:
      <input
        name="description"
        type="text"
        value=""
      />
    </label>
    <label>
      Ciudad:
      <input
        name="city"
        type="text"
        value=""
      />
    </label>
    <label>
      Departamento:
      <input
        name="states"
        type="text"
        value=""
      />
    </label>
    <label>
      Edades:
      <input
        name="age"
        placeholder="Ejemplo: 18-55"
        type="text"
        value=""
      />
    </label>
    <label>
      Fecha y Hora de Inicio:
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            type="text"
            value="12/11/2020 01:40"
          />
        </div>
      </div>
    </label>
    <label>
      Fecha y Hora de Fin:
      <div
        class="react-datepicker-wrapper"
      >
        <div
          class="react-datepicker__input-container"
        >
          <input
            class=""
            type="text"
            value="12/11/2020 01:40"
          />
        </div>
      </div>
    </label>
    <button
      class="sc-eCssSg gJtyAZ form-component"
      type="button"
    >
      Crear Propuestas
    </button>
    <input
      class="form-component"
      type="submit"
      value="Submit"
    />
  </form>
</div>
`;
